scipy/optimize/_zeros_py.py

```diff
@@ -1393,3 +1393,328 @@ def toms748(f, a, b, args=(), k=1,
+
+    funcalls = 0
+
+    # Ensure inputs are numpy arrays and converted to complex.
+    p0 = np.asarray(x0, dtype=complex)[()] * 1.0
+    p1 = np.asarray(x1, dtype=complex)[()] * 1.0
+    p2 = np.asarray(x2, dtype=complex)[()] * 1.0
+
+    converged = False
+    method = "muller"
+
+    if x0 == x1 or x0 == x2 or x1 == x2:
+        raise ValueError("x0, x1, and x2 must be different")
+
+    f0, f1, f2 = f(p0, *args), f(p1, *args), f(p2, *args)
+    funcalls += 3
+
+    # If one of the guesses is a root, return it immediately.
+    if np.isclose(f0, 0, rtol=rtol, atol=tol):
+        return _results_select(full_output, (x0, funcalls, 0, _ECONVERGED), method)
+    if np.isclose(f1, 0, rtol=rtol, atol=tol):
+        return _results_select(full_output, (x1, funcalls, 0, _ECONVERGED), method)
+    if np.isclose(f2, 0, rtol=rtol, atol=tol):
+        return _results_select(full_output, (x2, funcalls, 0, _ECONVERGED), method)
+
+    for itr in range(maxiter):
+        p1p0 = p1 - p0
+        p2p1 = p2 - p1
+        p0p2 = p0 - p2
+
+        # Gracefully handle division by zero.
+        # This can be avoided entirely by taking the zero value to be some epsilon > 0,
+        # but in my experience this leads to silent failure to converge that's
+        # difficult to diagnose.
+        if p1p0 == 0 or p2p1 == 0 or p0p2 == 0:
+            msg = "Division by zero has occured. Two initial values might be too close."
+            if disp:
+                msg += (
+                    " Failed to converge after %d iterations, value is %s."
+                    % (itr + 1, p2))
+                raise RuntimeError(msg)
+            warnings.warn(msg, RuntimeWarning, stacklevel=2)
+            return _results_select(
+                full_output, (p2, funcalls, itr + 1, _ECONVERR), method)
+
+        f0_f1_diff = (f0 - f1) / p1p0
+        f1_f2_diff = (f1 - f2) / p2p1
+        f0_f2_diff = (f0 - f2) / p0p2
+
+        w = f1_f2_diff + f0_f2_diff - f0_f1_diff
+
+        discriminant = np.sqrt(w**2 - 4 * f2 * f0_f2_diff)
+
+        # Chose maximal denominator for individual values.
+        denominator = (
+            w + discriminant
+            if np.abs(w + discriminant) > np.abs(w - discriminant)
+            else w - discriminant
+        )
+
+        h = -2 * f2 / denominator
+        p = p2 + h
+
+        converged = np.isclose(h, 0, rtol=rtol, atol=tol)
+
+        if converged:
+            return _results_select(
+                full_output, (p, funcalls, itr + 1, _ECONVERGED), method
+            )
+        
+        p0, p1, p2 = p1, p2, p
+
+        f0, f1, f2 = f1, f2, f(p2, *args)
+        funcalls += 1
+
+    if disp:
+        msg = "Failed to converge after %d iterations, value is %s." % (itr + 1, p2)
+        raise RuntimeError(msg)
+    
+    return _results_select(full_output, (p, funcalls, itr + 1, _ECONVERR), method)
+
+
+def _array_muller(f, x0, x1, x2, args, tol, rtol,
```

Comment:
Let's start with just the scalar version for simplicity since the vectorised version will end up living somewhere else